### DISCUSSION TITLE
Broken dag issue

### DISCUSSION THREAD

**Discussion by rohan472000** [apache/airflow#28661] [Q&A]
Broken DAG: [/opt/airflow/dags/dag.py] Traceback (most recent call last):
  File "/opt/airflow/dags/dag.py", line 7, in <module>
    from training import training
  File "/opt/airflow/dags/training.py", line 6, in <module>
    from joblib import dump
ModuleNotFoundError: No module named 'joblib'

But I have installed joblib already then why it is showing this error??

**Comment by rohan472000** [comment]
update from Airflow v2.1.1 and latest version, where you can specify the python lib(s) while configuring docker-compose.yml file.

_PIP_ADDITIONAL_REQUIREMENTS: ${_PIP_ADDITIONAL_REQUIREMENTS:- joblib==1.2.0}

You can find the more information related to above snippet below:

[docker-compose.yml](https://airflow.apache.org/docs/apache-airflow/stable/docker-compose.yaml)

[Environment variables supported by Docker Compose](https://airflow.apache.org/docs/apache-airflow/stable/howto/docker-compose/index.html) - Check bottom section.

And don't forget to restart the docker after this changes.

docker-compose restart

if changes do not reflect by above command then use below command:-

docker-compose up

This will install all the required python lib(S) along with Airflow services.

**Comment by rohan472000** [comment]
I got my error fixed with above procedures but if u have any other way to rectify it then do comment your answers.

**Reply by potiuk** [reply]
As mentioned in the note you mentioned. You should build your own image and extend it by adding the packages you need https://airflow.apache.org/docs/docker-stack/build.html#adding-new-pypi-packages-individually 

Using `_PIP_ADDITIONAL_REQUIREMENTS` is highly discouraged for anything but the quick iteration while debugging your installation (see the note for details).

**Reply by rohan472000** [reply]
> As mentioned in the note you mentioned. You should build your own image and extend it by adding the packages you need https://airflow.apache.org/docs/docker-stack/build.html#adding-new-pypi-packages-individually
> 
> Using `_PIP_ADDITIONAL_REQUIREMENTS` is highly discouraged for anything but the quick iteration while debugging your installation (see the note for details).

yes I tried by building my own image earlier but it didn't rectified my issue, so I went forward with _PIP_ADDITIONAL_REQUIREMENTS

**Reply by potiuk** [reply]
Possibly because you have not used it after you built it. Using _PIP_ADDITIONAL_REQUIREMENTS will hit you back much harder eventually. This all but certain.

**Reply by rohan472000** [reply]
yeah it could be a reason, let me try it from a fresh start.

**Reply by potiuk** [reply]
BTW. you don't even have to do much. Docker compose fully supports automatic building image with new dependencies and using it. See this comment in the dockerfile:

 https://github.com/apache/airflow/blob/main/docs/apache-airflow/howto/docker-compose/docker-compose.yaml#L46

```
  # In order to add custom dependencies or upgrade provider packages you can use your extended image.
  # Comment the image line, place your Dockerfile in the directory where you placed the docker-compose.yaml
  # and uncomment the "build" line below, Then run `docker-compose build` to build the images.
  image: ${AIRFLOW_IMAGE_NAME:-apache/airflow:|version|}
  # build: .
```

**Reply by potiuk** [reply]
It's all there, explained rather clearly.

**Reply by potiuk** [reply]
You can also run `docker compose up --build` as a shortcut if you do not want to run `docker compose build` separately.

**Reply by rohan472000** [reply]
> BTW. you don't even have to do much. Docker compose fully supports automatic building image with new dependencies and using it. See this comment in the dockerfile:
> 
> https://github.com/apache/airflow/blob/main/docs/apache-airflow/howto/docker-compose/docker-compose.yaml#L46
> 
> ```
>   # In order to add custom dependencies or upgrade provider packages you can use your extended image.
>   # Comment the image line, place your Dockerfile in the directory where you placed the docker-compose.yaml
>   # and uncomment the "build" line below, Then run `docker-compose build` to build the images.
>   image: ${AIRFLOW_IMAGE_NAME:-apache/airflow:|version|}
>   # build: .
> ```

@potiuk  Wow..it worked for me..I read the documentation of apache-airflow and this is the best practice to install additional packages.

**Reply by potiuk** [reply]
Reading docs helps, yep.
